### ISSUE TITLE
dok_matrix no longer allows update

### ISSUE THREAD

**Issue by amillb** [scipy/scipy#8338]
In previous versions, scipy allowed an `update` on `sparse.dok_matrix`. This seems to be no longer the case with v1.0.0.

With version 0.19.0, no error is thrown, and the matrix is updated accordingly.

The [docs](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.dok_matrix.html#scipy.sparse.dok_matrix) imply that the `update` method is still supported. If this isn't a bug, is there a suggested approach to update a `dok_matrix`?

### Reproducing code example:
```
from scipy import sparse
m = sparse.dok_matrix((10, 10))
m.update({(1,1):'a'})
```
### Error message:
```
NotImplementedErrorTraceback (most recent call last)
<ipython-input-15-b3e12160bd9e> in <module>()
      1 from scipy import sparse
      2 m = sparse.dok_matrix((10, 10))
----> 3 m.update({(1,1):'a'})
      4 

/Users/amb/Library/Enthought/Canopy/edm/envs/User/lib/python2.7/site-packages/scipy/sparse/dok.pyc in update(self, val)
    115     def update(self, val):
    116         # Prevent direct usage of update
--> 117         raise NotImplementedError("Direct modification to dok_matrix element "
    118                                   "is not allowed.")
    119 

NotImplementedError: Direct modification to dok_matrix element is not allowed.
```

### Scipy/Numpy/Python version information:
```
('1.0.0', '1.14.0', sys.version_info(major=2, minor=7, micro=13, releaselevel='final', serial=0))
```

**Comment by WarrenWeckesser** [comment]
The change was made in https://github.com/scipy/scipy/pull/7673

@perimosocordiae, can you comment on this?

> [...] is there a suggested approach to update a `dok_matrix`?

You can assign a value to a `dok_matrix`:

```
In [20]: m = sparse.dok_matrix((5, 10))

In [21]: m[1, 2] = -10.0

In [22]: m.A
Out[22]: 
array([[  0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.],
       [  0.,   0., -10.,   0.,   0.,   0.,   0.,   0.,   0.,   0.],
       [  0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.],
       [  0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.],
       [  0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.]])
```

**Comment by amillb** [comment]
> You can assign a value to a dok_matrix:

Thanks. However, the `update` function is convenient for assigning multiple values at once from a dictionary. I suppose that I could iterate through that dictionary, but that's more cumbersome.

**Comment by perimosocordiae** [comment]
The `update` method sets values directly into the `dict` that backs a `dok_matrix`, a consequence of subclassing `dict`. We removed access to it intentionally because it bypasses the sanity checking that normal assignment provides, and makes it easy to accidentally put the matrix into an inconsistent state.

The recommended way to bulk-add elements to a DOK matrix is either (a) use fancy indexing to assign like you would for a dense array/matrix, or (b) loop over your values and use scalar-index assignment.

If you really must have the speed of a raw `dict.update`, you can use the internal `dok_matrix._update` method, which bypasses the checks and just inserts whatever you give it.

**Comment by WarrenWeckesser** [comment]
Wouldn't it be possible to implement an `update` method that acted like `dict.update` but also did the appropriate sanity checking?

**Comment by perimosocordiae** [comment]
Sure, that seems reasonable.

**Comment by amillb** [comment]
It seems that the `dok_matrix._update()` function has now been removed.

Are there any plans to implement a new `update` function in line with the discussion above? This would have better performance than iterating over the dictionary to update the matrix one value at a time.

**Comment by dschult** [comment]
The `dok_matrix._update` function can be reproduced using `m._dict.update(d)`.  But as before, this does not check the input for whether it might corrupt the data structure of the underlying dict.

It's not clear to me whether implementing a new `update` function in line with the discussion above would actually make it any faster than iterating over the dictionary to update the matrix. It would presumably require a pass through the input dict to check that all keys are valid (2-tuples of integers within the right bounds) and then another pass to do the updating. (The two passes allow you to use Python's fast built-in dict.update on the underlying data structure.) While it is quite possible that using `dict.update` is faster than updating one at a time, the key will be whether we can make checking the incoming keys fast enough. Can you try timing a brute force version that checks that all keys are 2-tuples of integers between 0 and `m.shape` values, followed by `m._dict.update(d)`?  Then compare to looping over `d.items()` and assigning one at a time.

**Comment by amillb** [comment]
Yes, `dict_update` does seem faster. Updating 1000 random items in a `dok_matrix`:

- 2.16 ms ± 8.3 μs per loop for assigning one at a time (looping over `d.items()`)
- 407 μs ± 1.37 μs per loop using the `._dict.update()` method with the checks

The checks are:

```
 assert np.all([isinstance(k, tuple) and len(k)==2 for k in d.keys()])
 assert np.all([isinstance(k1, (int, np.integer)) and isinstance(k2, (int, np.integer)) and k1<m.shape[0] and k2<m.shape[1] and k1>=0 and k2>=0 for (k1, k2) in d.keys()])
```

**Comment by dschult** [comment]
This has been addressed with a public `update` method added in #23625.  :tada:
